### ISSUE TITLE
Introduce a high-level layer API

### ISSUE THREAD

**Issue by seanmor5** [elixir-nx/axon#2]
The next step after #1 is to implement higher-level constructs on top of the lower-level functional implementations. The goal of the higher-level API is to provide abstractions for building neural networks that:

* limit the overhead of writing a neural network from scratch
* are easy to understand, especially for beginners
* are flexible enough for more complicated architectures (ResNets, GANs, etc.)
* can be represented as composition of Nx functions for low-level JIT/AOT compilation OR can be represented as higher-level constructs for compilation using specific NN compilers (ONNX, TFLite, etc.)

For simplicity, we'll leave the discussion of efficient handling of network state to a later issue. This issue will only focus on the architecture/representation of a network.

## Axon Struct

We will introduce an `%Axon{}` struct that represents a constructed network. For now, the struct will have the following attributes:

* `:input` - input to this layer/model/etc., or in the case of a literal `input` layer some metadata
* `:shape` - shape of the layer's parameters, can be inferred from `:input`, we can also allow input layer shapes to have `nil` batch dimensions to represent arbitrary sized batches
* `:transformation` - how does this layer transform the input, can be an atom (like `:dense`) which resolves to an already implemented layer, or a numerical definition for arbitrarily complex transformations
* other - there's a lot of other metadata that should be included in some way: `:initializer` and initializer options, `:activation` and activation options, layer specific options, possibly constraints, callbacks, etc.

The struct would be build up with calls to high-level functions in the root Axon namespace. For example, MNIST:

```elixir
model =
  Axon.input({nil, 784})
  |> Axon.dense(128, activation: :relu)
  |> Axon.dense(10, activation: :softmax)
```

Then a model would be compiled, perhaps like:

```elixir
compiled_model = Axon.compile(model, options)
``` 

At a minimum `Axon.compile` initializers parameters and returns a compiled function using whatever backend the user specifies. It can also be abstracted away in some higher-level training logic, but that's discussion for another issue.

Structuring the network in this way makes arbitrary compilation easy, the API is simple and easy to understand, and flexible enough for complex models. For example, a GAN:

```elixir
generator =
   Axon.input({nil, 128})
   |> Axon.dense(256, activation: :tanh)
   |> Axon.dense(512, activation: :tanh)
   |> Axon.dense(784, activation: :tanh)
   |> Axon.reshape({28, 28})

discriminator =
   Axon.input({nil, 784})
   |> Axon.dense(128, activation: :relu)
   |> Axon.dense(1, activation: :sigmoid)

combined =
  generator
  |> Axon.compose(discriminator)
```

## Layers

High-level layers are tied directly to their functional implementations in `Axon.Layers`. Some of them have layer specific options which can be passed during layer creation.

## Combinators

We will use combinators similar to: https://thinc.ai/docs/api-layers/#combinators to represent more complex relationships between layers. At a minimum we'd have:

* `compose` - function composition
* `add` - adds layers
* `concat` - concats layers
* `residual` - residual output
* `parallel`/`split` - something to represent multiple-model outputs
